widget/widget.go

```diff
@@ -13,51 +15,95 @@ type BaseWidget struct {
 	position fyne.Position
 	Hidden   bool
 
-	impl fyne.Widget
+	impl         fyne.Widget
+	propertyLock sync.RWMutex
 }
 
 // ExtendBaseWidget is used by an extending widget to make use of BaseWidget functionality.
 func (w *BaseWidget) ExtendBaseWidget(wid fyne.Widget) {
+	w.propertyLock.Lock()
+	defer w.propertyLock.Unlock()
+
 	if w.impl != nil {
 		return
 	}
 
 	w.impl = wid
 }
 
+// ReadFields provides a guaranteed thread safe way to access widget fields.
+func (w *BaseWidget) ReadFields(f func()) {
+	w.propertyLock.RLock()
+	defer w.propertyLock.RUnlock()
+
+	f()
+}
+
+// SetFieldsAndRefresh helps to make changes to a widget that should be followed by a refresh.
+// This method is a guaranteed thread-safe way of directly manipulating widget fields.
+func (w *BaseWidget) SetFieldsAndRefresh(f func()) {
+	w.propertyLock.Lock()
+	f()
+	w.propertyLock.Unlock()
+
+	w.Refresh()
```

Comment:
What if the basewidget has been extended?
```suggestion
	if w.impl == nil {
		w.Refresh()
	} else {
		w.impl.Refresh()
	}
```

Reply:
I don't think that's how Refresh lookup works - the extended object might provide the Refresh call, and if not then basewidget will do the lookup.

Reply:
If impl does not define Refresh, then BaseWidget.Refresh will call Refresh on the impl's Renderer.
If impl does define Refresh, this will still call Refresh on the impl's Renderer

Reply:
Ah yes I see your point, thanks :)